Account for `altExp` in converting SCE to AnnData

allyhawkins opened the pull request:
Closes #351 

Here I am accounting for the presence of an `altExp` when converting SCE objects to `AnnData` objects. 

- I updated the existing Nextflow process to convert SCE to AnnData to optionally output a second `hdf5` file corresponding to the `altExp` object. 
 - In the Rscript responsible for converting SCE to AnnData, I included additional arguments where users can provide an optional `feature_name` and output file for feature data. When `feature_name` is specified, it extracts the corresponding `altExp` from the object and writes it to a separate AnnData object.
- All outputs from the export process are mixed together into a single channel. I updated the `groupTuple` option to account for up to 6 files keeping `remainder:true` so that the smaller tuples with only three files won't be discarded. 
- I tested this with both a CITE-seq library and an RNA-only library.

jashapiro commented on modules/export-anndata.nf:
```diff
@@ -41,9 +47,9 @@ workflow sce_to_anndata{
       // creates anndata channel with [library_id, unfiltered, filtered, processed]
-      anndata_ch = export_anndata.out
+      anndata_ch = export_anndata.out.rna_anndata.mix(export_anndata.out.feature_anndata)
         .map{ [it[0]["library_id"], it[0], it[1]] }
```
I haven't tested this, but nextflow does have a feature to use variable group sizes with `groupKey`. I believe the syntax should look something like this, followed by removing the `size:` parameter in groupTuple. I'm also using the lambda syntax here to keep the mapping process a bit more a bit more clear, I hope, by automatically unpacking the input tuple to a series of variables (`meta` and `hdf5file`). (I'm realizing this syntax might be useful elsewhere where keeping track of what is where in a channel might get confusing, rather than it[0] all the time).

```suggestion
        .map{ meta, hdf5_file -> tuple(
          groupKey(meta.library_id, meta.feature_type in ["adt", "cellhash"]? 6 : 3),
          meta, 
          hdf5_file)
        }
```

I just had this thought while waiting to be picked up for dinner, so I will do more review later!

jashapiro commented on modules/export-anndata.nf:
```diff
@@ -41,9 +47,12 @@ workflow sce_to_anndata{
+          groupKey(meta.library_id, meta.feature_type in ["adt", "cellhash"]? 6 : 3),
+          meta,
+          hdf5_file)
+        }
```
I think you still need a `groupTuple` here?
```suggestion
        }
        .groupTuple(by: 0, remainder: true)
```

jashapiro commented on bin/sce_to_anndata.R:
```diff
@@ -16,9 +16,20 @@ option_list <- list(
+    help = "Feature type. Must match the altExp name, if present."
+  ),
+  make_option(
+    opt_str = c("--output_feat_h5"),
```
Three extra characters seem worth it here
```suggestion
    opt_str = c("--output_feature_h5"),
```

jashapiro commented on bin/sce_to_anndata.R:
```diff
@@ -44,5 +55,29 @@ sce <- readr::read_rds(opt$input_sce_file)
+  }
+  
+  # check for output file 
+  if(!(stringr::str_ends(opt$output_feat_h5, ".hdf5|.h5"))){
```
following earlier suggestion
```suggestion
  if(!(stringr::str_ends(opt$output_feature_h5, ".hdf5|.h5"))){
```

jashapiro commented on bin/sce_to_anndata.R:
```diff
@@ -44,5 +55,29 @@ sce <- readr::read_rds(opt$input_sce_file)
+  # export altExp sce as anndata object
+  scpcaTools::sce_to_anndata(
+    alt_sce,
+    anndata_file = opt$output_feat_h5
```
following earlier suggestion
```suggestion
    anndata_file = opt$output_feature_h5
```

jashapiro commented on modules/export-anndata.nf:
```diff
@@ -8,19 +8,25 @@ process export_anndata{
-      hdf5_file = "${meta.library_id}_${file_type}.hdf5"
-
+      rna_hdf5_file = "${meta.library_id}_${file_type}.hdf5"
+      feature_hdf5_file = "${meta.library_id}_${file_type}_feature.hdf5"
```
Should we make the file name more informative?
```suggestion
      feature_hdf5_file = "${meta.library_id}_${file_type}_{meta.feature_type}.hdf5"
```

I guess we should also decide if we want to use `_rna.hdf5` for the RNA file name?

allyhawkins replied:
I thought about using RNA, but then that means every file is going to have `_rna` even in the absence of any additional ADT or cellhash data. But maybe that's not a bad thing. 
I was also trying to be consistent with the SCE filenames.

jashapiro commented on modules/export-anndata.nf:
```diff
@@ -8,19 +8,25 @@ process export_anndata{
-
+      rna_hdf5_file = "${meta.library_id}_${file_type}.hdf5"
+      feature_hdf5_file = "${meta.library_id}_${file_type}_feature.hdf5"
+      feature_present = meta.feature_type in ["adt", "cellhash"]
```
If the above suggestion fails when `meta.feature_type` is not defined, we could do:
```suggestion
      if (meta.feature_type in ["adt", "cellhash"]){
        feature_present = true
        feature_hdf5_file = "${meta.library_id}_${file_type}_{meta.feature_type}.hdf5"
      } else {
        feature_present = false
      }
```

jashapiro commented on modules/export-anndata.nf:
```diff
@@ -8,19 +8,25 @@ process export_anndata{
     stub:
-      hdf5_file = "${meta.library_id}_${file_type}.hdf5"
+      rna_hdf5_file = "${meta.library_id}_${file_type}.hdf5"
+      feature_hdf5_file = "${meta.library_id}_${file_type}_feature.hdf5"
```
This should probably match the logic and filenames above, whatever we end up deciding there.

jashapiro commented on modules/export-anndata.nf:
```diff
@@ -8,19 +8,25 @@ process export_anndata{
       tuple val(meta), path(sce_file), val(file_type)
     output:
-      tuple val(meta), path(hdf5_file)
+      tuple val(meta), path(rna_hdf5_file), emit: rna_anndata
+      tuple val(meta), path(feature_hdf5_file), emit: feature_anndata, optional:true
```
I'm wondering if we might want to handle this a bit differently so that we have all of the hdf5 files in a single object in the output.

If we use the fact that `path` allows for globs, we could make the output something like:
```
tuple val(meta), path("{${meta.library_id}_${file_type}*.hdf5")
```
I think that would catch both files. We then would need to handle the possibility of multiple files in downstream steps, but I don't think that would be too bad, and it would save us from having to deal with any of the weird `groupKey` and `groupTuple` stuff for this part of the workflow.

jashapiro reviewed (commented):
My review is a bit backward here, because I started with the `groupKey` stuff, but I think we can get rid of it all by using an output `path` glob rather than optional outputs, which seems more efficient and in keeping with what we might ultimately want. (In some ways it might obviate my desire for  using `muData` objects as well).

I'm hoping this can work for us, but let me know what problems you run into, as I am sure there will be some, as any time we try to use a new Nextflow feature!

allyhawkins commented:
Thanks for the helpful suggestions. I adjusted the outputs to use the path glob rather than optional outputs. I did not add in the `if` statement around checking for `meta.feature_type`. I tested it with a library with and without an ADT or hash library and it still works regardless of the `meta.feature_type` being present or not. 

Regarding the grouping of the output, I kept it in with some slight modifications. I still want to output a tuple with the `meta` and all of the `hdf5` files corresponding to that library. I imagine as we start making modifications with the `hdf5` files and modifying the `AnnData` objects in the subsequent processes that we may go back and change this.

I also added the `_rna.hdf5` to the RNA only files. I think because each of these files only holds a single modality, we should specify that. This should be ready for another look.

jashapiro commented on modules/export-anndata.nf:
```diff
@@ -42,8 +47,13 @@ workflow sce_to_anndata{
       anndata_ch = export_anndata.out
-        .map{ [it[0]["library_id"], it[0], it[1]] }
-        .groupTuple(by: 0, size: 3, remainder: true)
+        .map{ meta, hdf5_files -> tuple(
+          groupKey(meta.library_id, meta.feature_type in ["adt", "cellhash"]? 6 : 3),
+          meta,
+          hdf5_files.collect()
+        )}
+        .groupTuple(by: 0, remainder: true)
+        //.groupTuple(by: 0, size: meta.feature_type in ["adt", "cellhash"]? 6 : 3, remainder: true)
         .map{ [it[1][0]] +  it[2] }
```
I'm pretty sure all of this is redundant at this point? 

The output from the workflow should be `tuple(meta, path(hdf5_files))` so the later processing should not be required, as that is the format we already want, isn't it?
```suggestion

```

allyhawkins replied:
I don't think so. Because we will have: 
```
[[meta, unfiltered_rna, unfiltered_feature], 
 [ meta, filtered_rna, filtered_feature],
 [ meta, processed_rna, processed_feature]]
```
But what I want is: 
```
[meta, [unfiltered_rna, unfiltered_feat, filtered_rna, filtered_feat, processed_rna, processed_feat]]
```
So having the grouping should produce this output?

jashapiro replied:
Ah, you are correct! But I think we may want it to be a bit different from how you show it: 

```
[meta, [unfiltered_hdf5s, filtered_hdf5s, processed_hdf5s]]
```

Which is to say that the hdf5s will be in subgroups, which I think is how we probably want it. (I don't think we need the `.collect()`, but let me know if I am missing something else!

But with that, we can  still simplify a bit to: 
```suggestion
        .map{ meta, hdf5_files -> tuple(
          meta.library_id, 
          meta,
          hdf5_files
        )}
        .groupTuple(by: 0, size: 3, remainder: true)
        .map{ [it[1][0]] +  it[2] }
```

allyhawkins commented:
I tested the most recent additions, so this is ready for another round of review.

jashapiro reviewed (approved):
LGTM!
